server/src/test/java/io/druid/server/QueryResourceV3Test.java

```diff
@@ -0,0 +1,136 @@
+* regarding copyright ownership. Metamarkets licenses this file
+* to you under the Apache License, Version 2.0 (the
+* "License"); you may not use this file except in compliance
+* with the License. You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an
+* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+* KIND, either express or implied. See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*/
+
+package io.druid.server;
+
+import com.google.common.collect.ImmutableMap;
+import com.metamx.common.guava.Yielder;
+import io.druid.segment.TestHelper;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ */
+public class QueryResourceV3Test
+{
+  @Test
+  public void testYielderCloseOnSerializationSuccess() throws Exception
+  {
+    final AtomicBoolean isClosed = new AtomicBoolean(false);
+
+    Yielder yielder = new Yielder()
+    {
+      @Override
+      public Object get()
+      {
+        throw new RuntimeException();
+      }
+
+      @Override
+      public Yielder next(Object initValue)
+      {
+        throw new RuntimeException();
+      }
+
+      @Override
+      public boolean isDone()
+      {
+        return true;
+      }
+
+      @Override
+      public void close() throws IOException
+      {
+        isClosed.set(true);
+      }
+    };
+
+    Map<String, Object> response = ImmutableMap.of(
+        "result", yielder,
+        "context", ImmutableMap.of()
+    );
+
+    try {
+      QueryResourceV3.serializeResponseToOS(
+          TestHelper.getObjectMapper().writer(),
+          new ByteArrayOutputStream(),
+          response
+      );
+    } catch (Exception ex) {
+      Assert.fail();
+    }
+
+    Assert.assertTrue(isClosed.get());
+  }
+
+  @Test
+  public void testYielderCloseOnSerializationFailure() throws Exception
+  {
+    final AtomicBoolean isClosed = new AtomicBoolean(false);
+
+    Yielder yielder = new Yielder()
+    {
+      @Override
+      public Object get()
+      {
+        throw new RuntimeException();
+      }
+
+      @Override
+      public Yielder next(Object initValue)
+      {
+        throw new RuntimeException();
+      }
+
+      @Override
+      public boolean isDone()
+      {
+        throw new RuntimeException();
+      }
+
+      @Override
+      public void close() throws IOException
+      {
+        isClosed.set(true);
+      }
+    };
+
+    Map<String, Object> response = ImmutableMap.of(
+        "result", yielder,
+        "context", ImmutableMap.of()
+    );
+
+    try {
+      QueryResourceV3.serializeResponseToOS(
+          TestHelper.getObjectMapper().writer(),
+          new ByteArrayOutputStream(),
```

Comment:
The thing I was concerned about was if the serializing fails *before* Jackson gets to the Yielder. This test wouldn't verify that (although it does verify a separate equally important thing).

Could you also do a test where the Yielder is just a normal Yielder (perhaps from a ResourceClosingSequence), but the output stream fails on all calls to `write`? I think that'd be enough to smoke out the case I'm thinking of.